File: Longest Substring_Without_Repeating_Characters.java
/**
 * Problem: https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * 
 * The key to solve this problem:
 * 1st, using hashtable to keep the latest index of current character.
 * 2nd, If current character is new, then...
 *      If current character is not new, then...
 * Be clear about these two questions above, problem will be solved.
 */
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s==null)
            return 0;
        int start = 0;
        int max = 0;
        int[] index = new int[256];
        for(int i=0; i<256; i++) index[i] = -1;
        
        for(int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if(index[c] == -1) {
                index[c] = i;
            } else {
                max = Math.max(max, i - start);
                for(int j=start; j<i; j++) {
                    if(s.charAt(j) == c) {
                        start = j + 1;
                        break;
                    }
                    index[s.charAt(j)] = -1;
                }
            }
        }
        max = Math.max(max, s.length() - start);
        return max;
    }
}

File: Swift/longest-substring-without-repeating-characters.swift
/**
 * Problem Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/
 *
 * In Swift Language, String doesnt have a native support for subscript. So we convert the String to an Array of String, where each String contains a single character. This will save tons of time, since String manipulation is very expensive operation.
 * 
 * String.characters.count will take Linear time.
 *
 */

class Solution {
    func lengthOfLongestSubstring(_ ss: String) -> Int {
        let s = ss.characters.map({String($0)})
        if s.count == 0 {
            return 0
        }
        
        var dict = [String : Int]()
        var maxl = 0
        var start = 0
        
        for (i, c) in s.enumerated() {
            if let x = dict[c] {
                maxl = max(maxl, i - start)
                for j in start..<i {
                    if s[j] == c {
                        start = j + 1
                        break
                    }
                    dict[s[j]] = nil
                }
            } else {
                dict[c] = i
            }
        }
        maxl = max(maxl, s.count - start)
        return maxl
    }
}
